low-level/files/opt/seagate/sspl/setup/health_provider/storage.py

```diff
@@ -0,0 +1,153 @@
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    def get_health_info(self, rpath):
+        """
+        Fetch health information for given FRU
+        rpath: Resouce path (Example: hw>controllers)
+        """
+        info = {}
+        nodes = rpath.strip().split(">")
+        leaf_node, _ = self.get_node_details(nodes[-1])
+        if leaf_node == "storage":
+            info["last_updated"] = int(time.time())
+            for fru in self.storage_frus:
+                info.update({fru: self.storage_frus[fru]()})
+        else:
+            fru = None
+            fru_found = False
+            for node in nodes:
+                fru, _ = self.get_node_details(node)
+                if self.storage_frus.get(fru):
+                    fru_found = True
+                    info = self.storage_frus[fru]()
+                    break
+        if not fru_found:
+            raise HealthProviderError(
+                errno.EINVAL,
+                "Health provider doesn't have support for'{rpath}'.")
+        return info
+
+    def get_controllers_info(self):
+        """Update and return controller information in specific format"""
+        data = []
+        controllers = self.get_realstor_show_data("controllers")
+        for controller in controllers:
+            controller_dict = {
+              "uid": controller.get("durable-id"),
+              "fru": "true",
+              "last_updated": int(time.time()),
+              "health": {
+                "status": controller.get("health", "NA"),
+                "description": controller.get("description"),
+                "recommendation": controller.get("health-recommendation"),
+                "specifics": [
+                  {
+                    "serial-number": controller.get("serial-number", "NA"),
+                    "disks": controller.get("disks", "NA"),
+                    "virtual-disks": controller.get("virtual-disks", "NA"),
+                    "model": controller.get("model", "NA"),
+                    "part-number": controller.get("part-number", "NA"),
+                    "fw": controller.get("sc-fw", "NA"),
+                    "location": controller.get("position", "NA")
+                  }
+                ]
+              }
+            }
+            data.append(controller_dict)
+        return data
+
+    @staticmethod
+    def get_realstor_show_data(fru: str):
+        """Fetch fru information through webservice API"""
+        from framework.platforms.realstor.realstor_enclosure import (
+            singleton_realstorencl as ENCL)
+
+        fru_data = []
+        fru_uri_map = {
+            "controllers": ENCL.URI_CLIAPI_SHOWCONTROLLERS
+        }
+        url = ENCL.build_url(fru_uri_map.get(fru))
+        response = ENCL.ws_request(url, ENCL.ws.HTTP_GET)
+
+        if not response or response.status_code != ENCL.ws.HTTP_OK:
+            return []
```

Comment:
log error, to a resource map specific log file

Reply:
Ok, noted.